### PR TITLE
Constant optimizer fix

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libevmasm/ConstantOptimiser.cpp [hunk 1]
```diff
@@ -203,16 +203,21 @@ AssemblyItems ComputeMethod::findRepresentation(u256 const& _value)
 			u256 powerOfTwo = u256(1) << bits;
 			u256 upperPart = _value >> bits;
 			bigint lowerPart = _value & (powerOfTwo - 1);
-			if (abs(powerOfTwo - lowerPart) < lowerPart)
+			if (powerOfTwo - lowerPart < lowerPart)
```
Comment (axic) [inline libevmasm/ConstantOptimiser.cpp]: Can you keep the parentheses here for readability?

File: test/libsolidity/SolidityOptimizer.cpp [hunk 2]
```diff
@@ -1191,31 +1192,42 @@ BOOST_AUTO_TEST_CASE(clear_unreachable_code)
...
+				ra = m_a;
+				rb = m_b;
+				rc = m_c;
+				rd = m_d;
 			}
 		}
 	)";
-	compileBothVersions(sourceCode);
+	compileBothVersions(sourceCode, 0, "C", 1);
+	compareVersions("get()");
```
Comment (axic) [inline test/libsolidity/SolidityOptimizer.cpp:1226]: `get()` is called after `set()` too. Is that duplicate?

Comment (chriseth) [reply test/libsolidity/SolidityOptimizer.cpp:1226]: No, it is on purpose. It checks that the stored values actually match before and after.

File: docs/bugs.json [hunk 3]
```diff
@@ -1,5 +1,17 @@
 [
     {
+        "name": "ConstantOptimizerSubtraction",
+        "short": "In some situations, the optimizer replaces certain numbers in the code with routines that compute different numbers.",
```
Comment (axic) [inline docs/bugs.json]: This doesn't seem to follow the others below, those have `summary` and `description`, this one has `short` and `long`.

File: docs/bugs.json [hunk 4]
```diff
@@ -1,5 +1,17 @@
...
+        "long": "The optimizer tries to represent any number in the bytecode by routines that compute them with less gas. For some special numbers, an incorrect routine is generated. This could allow an attacker to e.g. trick victims about a specific amount of ether, or function calls to call different functions (or none at all).",
+        "link": "",
+        "introduced": "0.0.0",
+        "fixed": "0.4.11",
+        "severity": "low",
+        "check": {
```
Comment (axic) [inline docs/bugs.json]: This is called  `conditions` in the entries below.

**Review by axic — APPROVED** [review]

**Comment by pirapira** [comment]
I was asked to review this.  I might something interesting, so I add this to my pending list with priority M.  Currently, items with priority M spend seven months in my pending list.

**Pull request merged**